cpp/src/utilities/device_atomics.cuh

```diff
@@ -447,124 +416,72 @@ namespace detail {
-    int16_t atomicCASImpl(int16_t* address, int16_t compare, int16_t val)
-    {
-        using T = int16_t;
-        using T_int = unsigned int;
-
-        bool is_32_align = (reinterpret_cast<size_t>(address) & 2) ? false : true;
-        T_int * address_uint32 = reinterpret_cast<T_int *>
-            (reinterpret_cast<size_t>(address) - (is_32_align ? 0 : 2));
-
-        T_int old = *address_uint32;
-        T_int assumed;
-        T target_value;
-        uint16_t u_val = type_reinterpret<uint16_t, T>(val);
-
-        do {
-            assumed = old;
-            target_value = (is_32_align) ? T(old & 0xffff) : T(old >> 16);
-            if( target_value != compare ) break;
-
-            T_int new_value = (is_32_align) ? (old & 0xffff0000) | u_val
-                    : (old & 0xffff) | (T_int(u_val) << 16);
-            old = atomicCAS(address_uint32, assumed, new_value);
-        } while (assumed != old);
-
-        return target_value;
-    }
-
-    // -----------------------------------------------------------------------
-    // intermediate function resolve underlying data type
-    template <typename T, typename BinaryOp>
-    __forceinline__  __device__
-    T genericAtomicOperationUnderlyingType(
-        T* address, T const & update_value, BinaryOp op)
-    {
-        T ret = cudf::detail::genericAtomicOperationImpl<T, BinaryOp, sizeof(T)>()
-            (address, update_value, op);
-        return ret;
-    }
-
 } // namespace detail
 
 /** -------------------------------------------------------------------------*
- * @brief reads the `old` located at the `address` in global or shared memory,
+ * @brief compute atomic binary operation
+ * reads the `old` located at the `address` in global or shared memory,
  * computes 'BinaryOp'('old', 'update_value'),
  * and stores the result back to memory at the same address.
  * These three operations are performed in one atomic transaction.
  *
  * The supported cudf types for `genericAtomicOperation` are:
  * int8_t, int16_t, int32_t, int64_t, float, double,
- * cudf::date32, cudf::date64, cudf::timestamp, cudf::category.
+ * cudf::date32, cudf::date64, cudf::timestamp, cudf::category,
+ * cudf::nvstring_category, cudf::bool8
  *
  * @param[in] address The address of old value in global or shared memory
- * @param[in] val The value to be added
+ * @param[in] val The value to be computed
+ * @param[in] op  The binary operator used for compute
  *
  * @returns The old value at `address`
  * -------------------------------------------------------------------------**/
 template <typename T, typename BinaryOp>
 __forceinline__  __device__
 T genericAtomicOperation(T* address, T const & update_value, BinaryOp op)
 {
-    auto ret = cudf::detail::genericAtomicOperationUnderlyingType(
-         &cudf::detail::unwrap(*address),
-         cudf::detail::unwrap(update_value),
-         op);
+    auto ret=  cudf::detail::genericAtomicOperationImpl<T, BinaryOp, sizeof(T)>{}
```

Comment:
I don't understand how the sizeof(T) works here. Is it not enough to specialize `genericAtomicOperationImpl` for the typename `T`? Do you need to also specialize on its size? Won't they always be one-to-one?

Reply:
I beleiece it should work since the caller `genericAtomicOperator` is a force inlined template function.
`sizeof(T)` would be evaluated when `genericAtomicOperator` is called by a cudf data type, otherwise the compilation will be failed.

Btw, I realized 'W genericAtomicOperator(W)' does not have `__forceinline__  __device__` qualifier, I will add.

Reply:
I guess what I don't understand is why you *need* sizeof(T) -- why do you need the size of the type T in bytes as a template parameter if you already have `typename T` as a template parameter. Is it possible for a single type `T` to have two different sizes? I don't think it is, so the size parameter seems redundant.

Reply:
Aha, you mean sizeof(T) is redundant.
Yes, It would be possible to move sizeof(T) into specialization pattern of `genericAtomicOperationImpl`.
I will try.